### ISSUE TITLE
How to improve the precision of masks for small objects？

### ISSUE THREAD

**Issue by panp4n** [ultralytics/ultralytics#24084]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

How to improve the precision of masks？

### Additional

I have tried two schemes:
1. Input size 640×640, change mask_ratio to 1, and modified the proto output size to 1×32×640×640. The effect is good for large objects, but very poor for slender object segmentation,like [#20200 ](https://github.com/ultralytics/ultralytics/issues/20200).
2. Input size 1280×1280 with mask_ratio = 4. The mask precision improves, but there is still jaggedness, and this input size will cut off the object mask, resulting in a shorter segment, same as [#22687 ](https://github.com/ultralytics/ultralytics/issues/22687).

How can I obtain a higher-precision mask?

**Comment by Y-T-G** [comment]
What model are you using?

**Comment by panp4n** [comment]
> What model are you using?

YOLOv8、YOLO11

**Comment by glenn-jocher** [comment]
Thanks, that helps — on YOLOv8/YOLO11 this is mostly a segmentation-head limitation: masks are reconstructed from prototypes and then cropped to the predicted box in [the segmentation predictor](https://docs.ultralytics.com/reference/models/yolo/segment/predict/) and [mask ops](https://docs.ultralytics.com/reference/utils/ops/), so thin/small objects can look jagged or get shortened when the box is slightly tight. The best practical gains are usually a larger `-seg` model, higher `imgsz`, tiling/cropping so small objects occupy more pixels, and `retina_masks=True` for cleaner inference output; if you’re testing new runs, please also compare with [Ultralytics YOLO26](https://docs.ultralytics.com/models/yolo26/). If you can share a small image+label sample, we can check whether your case is mainly box-limited or data-limited.

**Comment by panp4n** [comment]
> Thanks, that helps — on YOLOv8/YOLO11 this is mostly a segmentation-head limitation: masks are reconstructed from prototypes and then cropped to the predicted box in [the segmentation predictor](https://docs.ultralytics.com/reference/models/yolo/segment/predict/) and [mask ops](https://docs.ultralytics.com/reference/utils/ops/), so thin/small objects can look jagged or get shortened when the box is slightly tight. The best practical gains are usually a larger `-seg` model, higher `imgsz`, tiling/cropping so small objects occupy more pixels, and `retina_masks=True` for cleaner inference output; if you’re testing new runs, please also compare with [Ultralytics YOLO26](https://docs.ultralytics.com/models/yolo26/). If you can share a small image+label sample, we can check whether your case is mainly box-limited or data-limited.

Thanks for your reply. This is  a small dataset for testing.
https://drive.google.com/file/d/147Og7seLXL5BSKiwGTM67RZIs4hhMFCC/view?usp=drive_link

**Comment by glenn-jocher** [comment]
Thanks for sharing it — for very thin objects, if your downstream pipeline can use polygons, `results[0].masks.xy` from the [Masks reference](https://docs.ultralytics.com/reference/engine/results/) usually gives a cleaner contour than the raw bitmap mask. Please also attach 1–2 representative image/label pairs directly in the issue and paste the exact `train` and `predict` args so we can reproduce whether the shortening is coming from box clipping or segment-head resolution.

**Comment by panp4n** [comment]
> Thanks for sharing it — for very thin objects, if your downstream pipeline can use polygons, `results[0].masks.xy` from the [Masks reference](https://docs.ultralytics.com/reference/engine/results/) usually gives a cleaner contour than the raw bitmap mask. Please also attach 1–2 representative image/label pairs directly in the issue and paste the exact `train` and `predict` args so we can reproduce whether the shortening is coming from box clipping or segment-head resolution.

Thanks for your advice. These are all args:
task: segment
mode: train
model: yolo11s-seg.pt
data: Seg_Data.yaml
epochs: 150
time: null
patience: 150
batch: -1
imgsz: 640
save: true
save_period: 10
cache: false
device: null
workers: 1
project: null
name: model_weight
exist_ok: true
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
freeze: None
multi_scale: false
overlap_mask: false
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
save_hybrid: false
conf: 0.5
iou: 0.7
max_det: 300
half: false
dnn: false
plots: false
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: true
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: 1
format: onnx
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: 12
workspace: 4
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 70.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 12.0
kobj: 1.0
label_smoothing: 0.0
nbs: 64
hsv_h: 0.0
hsv_s: 0.5
hsv_v: 0.5
degrees: 43.0
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.5
fliplr: 0.5
mosaic: 0.0
mixup: 0.0
copy_paste: 0.0
auto_augment: randaugment
erasing: 0.0
crop_fraction: 1.0
cfg: null
tracker: botsort.yaml

I use Openvino for inference, so I can only get masks from  the raw bitmap mask.

**Comment by panp4n** [comment]
> Thanks for sharing it — for very thin objects, if your downstream pipeline can use polygons, `results[0].masks.xy` from the [Masks reference](https://docs.ultralytics.com/reference/engine/results/) usually gives a cleaner contour than the raw bitmap mask. Please also attach 1–2 representative image/label pairs directly in the issue and paste the exact `train` and `predict` args so we can reproduce whether the shortening is coming from box clipping or segment-head resolution.

<img width="747" height="603" alt="Image" src="https://github.com/user-attachments/assets/f5131186-6305-4374-bcdb-6332cd38ef3d" />

**Comment by glenn-jocher** [comment]
Thanks, the args and screenshot help — with `retina_masks=True` already on, this looks more like the normal [segmentation postprocess](https://docs.ultralytics.com/reference/models/yolo/segment/predict/) / [mask ops](https://docs.ultralytics.com/reference/utils/ops/) limit for thin objects, where masks are thresholded and cropped to the box, than an OpenVINO-only issue. I’d next try `imgsz=1280`, `mask_ratio=1`, much lighter geometry augmentation (`degrees`, `scale`, `translate`, `flipud`), and a smaller `warmup_epochs`, then compare the same image from the `.pt` model and the OpenVINO export; if both outputs match, please paste that result with `yolo checks` so we can reproduce from there.

**Comment by panp4n** [comment]
Thank you. I tried `imgsz=1280`, `mask_ratio=1` and trained to an overfitted state. The mask accuracy has indeed improved slightly, but the improvement is not significant, and the target cutoff problem will occur.

Inference result of one image from the training set:
<img width="1920" height="1200" alt="Image" src="https://github.com/user-attachments/assets/a76edbdc-13c8-4585-99df-7776b5ea0262" />

Are there any other methods to optimize mask accuracy, such as modifying the structure of the segmentation module? I don't care much about inference time; I focus more on segmentation accuracy.

**Comment by Y-T-G** [comment]
@panp4n Do you only have 1 class?

**Comment by panp4n** [comment]
> [@panp4n](https://github.com/panp4n) Do you only have 1 class?

Yes, just one class.

**Comment by Y-T-G** [comment]
You can try this to increase mask output size:

https://github.com/ultralytics/ultralytics/issues/20200#issuecomment-3360487909

**Comment by panp4n** [comment]
> ### Search before asking
> * [x]  I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.
> 
> ### Question
> How to improve the precision of masks？
> 
> ### Additional
> I have tried two schemes:
> 
> 1. Input size 640×640, change mask_ratio to 1, and modified the proto output size to 1×32×640×640. The effect is good for large objects, but very poor for slender object segmentation,like [#20200 ](https://github.com/ultralytics/ultralytics/issues/20200).
> 2. Input size 1280×1280 with mask_ratio = 4. The mask precision improves, but there is still jaggedness, and this input size will cut off the object mask, resulting in a shorter segment, same as [#22687 ](https://github.com/ultralytics/ultralytics/issues/22687).
> 
> How can I obtain a higher-precision mask?

I tried it right from the start. The mask boundaries for large objects look better, but for medium and small objects, they have become worse—even the masks are not aligned with the targets at all.

**Comment by Y-T-G** [comment]
@panp4n You can try training YOLO26-seg

**Comment by panp4n** [comment]
> [@panp4n](https://github.com/panp4n) You can try training YOLO26-seg

I've tried it, but there was no improvement. Are there any enhancements to the YOLO26 segmentation module?

**Comment by Y-T-G** [comment]
It uses improved loss for better mask quality
